File: src/main/java/edu/neu/coe/info6205/bqs/Element.java
/*
 * Copyright (c) 2017. Phasmid Software
 */

package edu.neu.coe.info6205.bqs;

import java.util.Objects;

public class Element<Item> {
    Element(Item x, Element<Item> n) {
        item = x;
        next = n;
    }

    Element(Item x) {
        this(x, null);
    }

    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof Element)) return false;
        Element<?> element = (Element<?>) o;
        return item.equals(element.item) && Objects.equals(next, element.next);
    }

    public int hashCode() {
        return Objects.hash(item, next);
    }

    final Item item;
    Element<Item> next;

    public String toString() {
        return item + (next == null ? " (last)" : "");
    }
}

File: Python/src/bqs/element.py
from typing import Generic

from util.generic_type import T


class Element(Generic[T]):

    def __init__(self, item, _next=None):
        self.item = item
        self.next = _next

    # Return a string representation self.
    def __str__(self):
        return f"{self.item}{' (last)' if self.next is None else ''}"
